Public/Src/Cache/ContentStore/Distributed/NuCache/RocksDbContentLocationDatabase.cs

```diff
@@ -90,13 +88,28 @@ public RocksDbContentLocationDatabase(IClock clock, RocksDbContentLocationDataba
         /// <inheritdoc />
         protected override Task<BoolResult> ShutdownCoreAsync(OperationContext context)
         {
+            lock (ShutdownLock)
+            {
+                _compactionTimer?.Dispose();
+            }
+
             _keyValueStore?.Dispose();
+
             return base.ShutdownCoreAsync(context);
         }
 
         /// <inheritdoc />
         protected override BoolResult InitializeCore(OperationContext context)
         {
+            if (_configuration.FullRangeCompactionInterval != Timeout.InfiniteTimeSpan)
+            {
+                _compactionTimer = new Timer(
+                    _ => FullRangeCompaction(context),
+                    null,
+                    IsGarbageCollectionEnabled ? _configuration.FullRangeCompactionInterval : Timeout.InfiniteTimeSpan,

```

Comment:
This looks strange to me, why we need to instantiate timer if gc is disabled? In this case we'll create a timer that will never be fired.

Reply:
This is exactly the same logic we use for all other timers. The reason for it is that a machine could be promoted to master, in which case it needs to start using the timer.

---
In reply to: [313643144](https://github.com/microsoft/BuildXL/pull/731#discussion_r313643144) [](ancestors = 313643144)

Reply:
This is not entirely true. In some cases we actually check that a feature is disable as well:

if (_configuration.ContentCacheEnabled && _configuration.CacheFlushingMaximumInterval != Timeout.InfiniteTimeSpan)
            {
  /// timer creation
}


---
In reply to: [313644959](https://github.com/microsoft/BuildXL/pull/731#discussion_r313644959) [](ancestors = 313644959,313643144)

Reply:
And the logic is actually pretty inconsistent.

I would say that we should follow one pattern or another. But right now it is not clear why one configuration option is more important than another.

---
In reply to: [313646774](https://github.com/microsoft/BuildXL/pull/731#discussion_r313646774) [](ancestors = 313646774,313644959,313643144)

Reply:
In this particular piece of code, the same thing happens: if the full range compaction interval is infinite, it will automatically be disabled. The end result is equivalent to disabling the feature, and is the same thing we do with garbage collection.

---
In reply to: [313646774](https://github.com/microsoft/BuildXL/pull/731#discussion_r313646774) [](ancestors = 313646774,313644959,313643144)